RStudio IE and IE Framework Enhancements

hexylena opened the pull request:
Includes the entire history of the RStudio IE.

Depends on
- [x]  merging #229 **first**.
- [x] #239

bgruening committed:
Copied readme from Bjoern's repo

Commit:
Added licence/config file

bgruening committed:
Copied from bgruening/galaxy-ipython

Commit:
Added empty default Rproject file

Commit:
Initial version stuck to port 7777

Commit:
More progress

Autologin may not be possible but we'll certainly try for it. This has the container come up and
copies every dataset into the R container. No magic R/Galaxy sugar yet.

bgruening committed:
Update README.md

Commit:
Removed quotes

Commit:
Template updates

Commit:
Moved to IE framework

Commit:
More progress towards new framework

Commit:
Mostly working, CORS issues

Commit:
Merge branch 'ie-framework'

Commit:
Wait on pubkey to exist before launching

Commit:
Corrected AJAX request call

Commit:
Updated port for nginx proxy, added more error catching

Commit:
Updated URLs

Commit:
Moved docker container to subpath

Commit:
Extracted files from official rstudio github

They were different than files pulled from Tom Wu's personal webpage. As such, took apart the
following file:
https://github.com/rstudio/rstudio/blob/44b1b2815b76013d5a2ba6e9a203ab52be7e5d9e/src/gwt/tools/rsa.js
and extracted constituent parts.

Commit:
Prevent multiple requests from fufilling by timeing out

Commit:
Move encryption to JS only in rstudio.js

Much better than waiting on reading the file from OS.

Commit:
Merge pull request #1 from erasche/develop

Develop

Commit:
Last ditch JS attempts

Also threw in all of the wu/rstudio originals

Commit:
Remove IE specific text

Commit:
Merge pull request #2 from erasche/develop

Develop

Commit:
Updated the readme

Commit:
Fixed URLs

Commit:
Change to enable apache_urls usage

Commit:
Added installation instructions

Commit:
Merge branch 'master' of github.com:erasche/galaxy-rstudio

Commit:
Fixed login issue

Not sure I'm entirely fond of this solution, however it functions perfectly

Commit:
Mostly translated to new framework

Commit:
Removed docs to align with ipython

Commit:
Removed picture

Commit:
Fixed image name

jmchilton committed:
More fixes for moving interactive_environments into Galaxy.

Commit:
Merge pull request #3 from jmchilton/new-viz-framework

More fixes for moving IE framework into Galaxy.

Commit:
Temporary removal of non-root IE fix

RStudio does not (yet) support running as an unprivleged user. The patch we applied to handle that
was applied globally, resulting in issues for IEs that are not yet patched for root issues

Commit:
Fixes for use within new <interactive_environment>

Numerous fixes and bugs introduced here ;)

 - Does not work with @jmchilton's proxy
 - Messily handles re-writing of passwords, this is a bug in Galaxy-IE
   that needs to be patched by @erasche
 - Hacks out non-root container setting, as rstudio currently must run
   as root (PAM authentication)

Commit:
Merge branch 'new-viz-framework'

Commit:
Updated doctype

Commit:
Proxy seems to be working now?

Commit:
Removed warning

Commit:
Body was inheriting CSS from galaxy and havinga  margin

Commit:
Merge pull request #4 from erasche/new-viz-framework

New viz framework

Commit:
Automatically restore .RData files from history

Commit:
Pass CORS_ORIGIN to NB, make sure URLs are correct

Commit:
Set to port 80 for nginx

Commit:
Merge remote-tracking branch 'rstudio/master' into rstudio

Commit:
Migrated to correct location

jmchilton commented:
Can we label this one WIP also? We probably should add an RData datatype as part of this.

hexylena commented:
@jmchilton sure, that's fine.

jmchilton commented on config/plugins/interactive_environments/rstudio/templates/rstudio.mako:
```diff
@@ -0,0 +1,84 @@
+notebook_login_url =  ie_request.url_template('${PROXY_URL}/rstudio/${PORT}/auth-do-sign-in')
+
+# Did the user give us an RData file?
+if hda.datatype.__class__.__name__ == "RData":
```
What do you think about subtyping RData for this purpose (RStudio IE workspace)? Does that make sense - is there special stuff in the environment - or is one workspace the same as another?

hexylena replied:
Huh. That's actually a good question, I hadn't given that any thought.

`.RData` files should just be .gz of data structures loaded in the R session. I don't know think anything is included in the IE that wouldn't be able to be used outside of the IE, so I'd lean towards keeping it as-is, just raw RData files. I'll double check this tomorrow, that the .RData files are completely functional inside and out. If they are, I'd say no reason to impose any extraneous limits.

hexylena replied:
Okay, doesn't look like there's a need to subclass anything. Let's just go with this as-is, whenever you're happy with the PR.

Commit:
Pin to :dev

jmchilton commented on config/plugins/interactive_environments/rstudio/config/rstudio.ini.sample:
```diff
@@ -0,0 +1,11 @@
+[docker]
+command = docker
+image = erasche/docker-rstudio-notebook:dev
+
```
@erasche I don't think this tag exists (https://registry.hub.docker.com/u/erasche/docker-rstudio-notebook/tags/manage/) - it seems to for IPython (https://registry.hub.docker.com/u/bgruening/docker-ipython-notebook/tags/manage/). Can you mimic @bgruening's IPython setup or switch this to `:latest`?

hexylena replied:
My intention was to mirror @bgruening's. Whoops. Apparently I didn't turn on the `:dev` tag. Image should be building now. **Thank you** for catching that.

hexylena replied:
https://registry.hub.docker.com/u/erasche/docker-rstudio-notebook/builds_history/41411/

jmchilton commented:
So #229 is listed as prerequisite but this PR uses `write_conf_file` which is renamed as part of that.

```
galaxy.webapps.galaxy.controllers.visualization ERROR 2015-05-11 02:33:55,781 error rendering visualization (rstudio): 'InteractiveEnviornmentRequest' object has no attribute 'write_conf_file'
Traceback (most recent call last):
  File "/opt/galaxy/galaxy-app/lib/galaxy/webapps/galaxy/controllers/visualization.py", line 732, in render
    return plugin.render( trans=trans, embedded=embedded, **kwargs )
  File "/opt/galaxy/galaxy-app/lib/galaxy/visualization/plugins/plugin.py", line 139, in render
    return self._render( render_vars, trans=trans, embedded=embedded )
  File "/opt/galaxy/galaxy-app/lib/galaxy/visualization/plugins/plugin.py", line 280, in _render
    return trans.fill_template( template_filename, template_lookup=self.template_lookup, **render_vars )
  File "/opt/galaxy/galaxy-app/lib/galaxy/web/framework/webapp.py", line 810, in fill_template
    return self.fill_template_mako( filename, **kwargs )
  File "/opt/galaxy/galaxy-app/lib/galaxy/web/framework/webapp.py", line 825, in fill_template_mako
    return template.render( **data )
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/template.py", line 296, in render
    return runtime._render(self, self.callable_, args, data)
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/runtime.py", line 660, in _render
    **_kwargs_for_callable(callable_, data))
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/runtime.py", line 692, in _render_context
    _exec_template(inherit, lclcontext, args=args, kwargs=kwargs)
  File "/opt/galaxy/galaxy-app/eggs/Mako-0.4.1-py2.7.egg/mako/runtime.py", line 718, in _exec_template
    callable_(context, *args, **kwargs)
  File "/opt/galaxy/galaxy-app/database/template_cache/rstudio.mako.py", line 60, in render_body
    ie_request.write_conf_file(temp_dir, {'notebook_username': 'galaxy', 'notebook_password': PASSWORD,
AttributeError: 'InteractiveEnviornmentRequest' object has no attribute 'write_conf_file'
```

Thoughts?

hexylena commented:
This shjouldn't be using write_conf_file, crapsticks. I switched everything in RStudio over to ENV... I must not have pushed that commit. So sorry @jmchilton!

Commit:
Updates to take advantage of #229 infrastructure

hexylena committed:
Merge pull request #2 from erasche/rstudio2

Updates to take advantage of #229 infrastructure

hexylena commented:
@jmchilton pushed the commits. Sorry about that. Doing some more testing now just in case I missed anything...

hexylena commented:
@jmchilton okay, this was a less tested PR and definitely more WIP (mostly because I haven't quite figured out the waltz to developing on one branch while using commits from another that don't get included in the history), so I'm going to say "this is still WIP" and I'll get everything 100% tomorrow morning. Sorry about the bugs, thanks again for the code review and sanity checks, I really appreciate it!

jmchilton commented:
No problem at all - thank you for the fixes and for working this - this is really exciting functionality!

If you are going to work on this more - let me know when you think it is ready to go - I'll work on testing #229 by itself instead in the mean time.

hexylena commented:
Okay, I'll ping you again when it's out of WIP.

Commit:
Update RStudio to new IE infrastructure

Commit:
Update RStudio to new IE infrastructure

hexylena commented:
@jmchilton no longer WIP, if you'd like to test it :)

jmchilton committed:
Merge remote-tracking branch 'jmchilton/dev' into rstudio

jmchilton committed:
Expose and cache raw IE password to simplify RStudio plugin.

jmchilton committed:
Introduce new IE launch abstraction.

In addition to reducing duplication between IPython and RStudio - it also shields them from low-level details and creates an interface we could more reasonably generalize to something like qsubing the docker commands or using Docker Swarm.

jmchilton committed:
IEs: Move temp dir handling into ie_request to simplify interface.

jmchilton committed:
IE: Move IPython-specific password hashing out of framework into plugin.

See dicussion here - https://github.com/erasche/galaxy/commit/23459f318507255a2bb1d1fb647e1581bd2090e1.

hexylena committed:
Merge pull request #4 from jmchilton/rstudio

More IE Enhancements [WIP]

Commit:
Pin to :dev

CC @jmchilton @bgruening, pending the next Galaxy release, I've pinned both
images to their :dev tags.

This commit is really backing out a previous commit pinning to :15.05
which was the wrong thing to do as of #229

jmchilton committed:
Fix problems introduced in IE refactorings.

jmchilton commented:
Wonderful - works for me now! Thanks a million @erasche!

jmchilton commented:
Mind if I update the title to reflect that there are IE framework enhancements in there also?

hexylena commented:
Go for it!

bgruening commented:
Release time! This is exciting!
